release-23.1: roachprod: avoid adding new keys to project metadata when creating clusters

renatolabs opened the pull request:
Backport:
  * 1/1 commits from "roachprod: avoid adding new keys to project metadata when creating clusters" (#119106)
  * 1/1 commits from "roachprod: search ssh directory for keys" (#120251)

Please see individual PRs for details.

/cc @cockroachdb/release

Release justification: test only changes.

Commit:
roachprod: avoid adding new keys to project metadata when creating clusters

Roachprod uses GCP's 'project metadata' feature[^1] to manage SSH
keys. This is useful to make sure every cluster created by roachprod,
regardless of the cloud used, gets the same set of keys in the "shared
user" (`ubuntu`) account.

However, roachprod also calls `gcloud config-ssh` when creating new
clusters. This command, among other things, adds the caller's public
key to the project metadata.

TeamCity agents (where a lot of roachprod clusters are created in the
process of running roachtests) run in ephemeral environments with
public keys that are only valid during that run. As a consequence,
these ephemeral keys keep being added to the project metadata, making
us hit the storage limitation of 256KB for these keys.

This limit should be more than enough for our current usage. At the
time of writing, out of the 447 public keys stored in the project
metadata, only 50 are *not* from `agent` or `roach` (usernames
associated with TeamCity agents).

In this commit, we change how cluster provisioning in GCP works:
instead of relying on `config-ssh` to provide access to the newly
created VM, we include the caller's public key in the shared user's
`authorized_keys` right on startup. As a result, TeamCity agents can
run as usual (since creating a public key is already part of their
scripts -- it's required by the AWS provider) while *not* adding any
keys to the project metadata. This should allow the GCP project
metadata to be used exclusively for employee keys that should exist in
every cluster.

Informs: #90104

Release note: None

[^1]: https://cloud.google.com/compute/docs/connect/add-ssh-keys

ajwerner committed:
roachprod: search ssh directory for keys

Before this change, roachprod required that an id_rsa.pub file exist in
the user's $HOME/.ssh directory. These days folks use other types of
keys like ed25519. The github docs these days in fact explicitly tell
you to use ed25519 [1]. This patch now searches for keys the same way
that the openssh client does. It searches based on this list:

 * id_rsa
 * id_ecdsa
 * id_ecdsa_sk
 * id_ed25519
 * id_ed25519_sk
 * id_dsa

[1]: https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent

Epic: None

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/120444)
<!-- Reviewable:end -->

srosenberg commented:
```
compilepkg: missing strict dependencies:
    /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1814/execroot/com_github_cockroachdb_cockroach/pkg/roachprod/config/config.go: import of "slices"
```

Commit:
roachprod: remove `slices` dependency

This is a release-23.1 and below commit that is meant to remove the
`slices` dependency, unavailable in older releases.

Epic: none

Release note: None

renatolabs commented:
> ```
> compilepkg: missing strict dependencies:
>     /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1814/execroot/com_github_cockroachdb_cockroach/pkg/roachprod/config/config.go: import of "slices"
> ```

Fixed with a separate commit to be included only in 23.1-* and 22.2.

renatolabs commented:
[A build on this branch](https://teamcity.cockroachdb.com/viewLog.html?buildId=14395268&tab=queuedBuildOverviewTab) ran without problems.

renatolabs commented:
TFTR!
